### PR TITLE
CRISPR for Base Gene editing

### PR DESCRIPTION
**Pull request by read-0nly**
## About The Pull Request

Second try, hopefully with less spam

Changes DnaConsole.DM and tgui to add a new function, CRISPR editing. This allows the overwriting of a solved gene with another known gene by use of a "CRISPR string" representing the full transmutation required. It thus requires manually transcribing the known genes to create a string that replaces that specific gene with the specific required gene. Follows the idea of how [CRISPR-Cas9](https://en.wikipedia.org/wiki/CRISPR_gene_editing#Genome_engineering) currently needs both what you want to replace and want to replace it with.

So you write down the full sequence for both genes, then combine them by the lead of each pair like so:
```
 GC CG CG CG GC GC TA TA CG AT TA AT AT GC GC TA	< Replacement Sequence
 |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |
AG GC CC AC AG AG CT CT TC AA GT TA AA AG TG AT   	< CRISPR string
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  | 
AT GC CG AT AT AT CG CG TA AT GC TA AT AT TA AT   	< Old sequence
```

This would give you the gene string AGGCCCACAGAGCTCTTCAAGTTAAAAGTGAT, which you would then feed in when using CRISPR on a live human's mutations. The difficulty and threshold of knowledge required should hopefully limit abuse, but it's not over - If you mess it up, instead you get a scrambled ACIDFLESH with no guides. To recover it, you need to manually set it to the proper sequence, then overwrite ACIDFLESH with the desired sequence. If you mess up on more than one gene, they'll collapse into ACIDFLESH, so you could lose a permanent slot with it. And then to add to the risk, since we're messing with viruses and DNA, you also get infected with a random virus every time you use it.

Requires the console be fed an activator that was used on a subject infected with a virus with either Dormant DNA Activator or Viral Evolutionary Acceleration. The idea is that the virus sample is then used to drive the CRISPR action. 

Basically a re-submit of [this PR](https://github.com/tgstation/tgstation/pull/53097) from a fresh fork

## Why It's Good For The Game

Adds a way to sidestep instability by editing base genes, but with a complexity and risk high enough to dissuade casual grief. Occasional project of turning someone into a supersoldier, if you're willing to take the trouble. 

Forces both Viro and Geneticist to work together, both very isolated roles.

## Changelog
:cl:
add: Adds CRISPR - a gene editing method that lets you overwrite base mutations
/:cl:

### CODE REVIEW

**Comment by read-0nly** [comment]
Charge counter on injectors isn't working, debugging now

**Comment by read-0nly** [comment]
Testing suggests it's ready to go, behaves as expected

**Comment by read-0nly** [comment]
Tested with uninfected subject, infected subject without the required symptoms, infected subject with right symptoms. Only provides charge on the latter. Suffix applies as expected. Console consumes and increments the charge, the function only works if there's a charge available and decrements the count

File: code/game/machinery/computer/dna_console.dm [hunk 1]
```diff
@@ -549,6 +554,135 @@
...
+		//   "disk" - From inserted diskette
+		if("crispr")
+			// GUARD CHECK - Can we genetically modify the occupant? Includes scanner
+			//  operational guard checks.
+			if(!can_modify_occupant())
+				return
+
+			// GUARD CHECK - Have we somehow cheekily swapped occupants? This is
+			//  unexpected.
+			if(!(scanner_occupant == connected_scanner.occupant))
```
Comment (stylemistake) [inline code/game/machinery/computer/dna_console.dm:578]:
```suggestion
			if(scanner_occupant != connected_scanner.occupant)
```

File: code/game/machinery/computer/dna_console.dm [hunk 2]
```diff
@@ -549,6 +554,135 @@
...
+			var/bref = params["mutref"]
+
+			//Valid gene-pairs
+			var/atstr = "AT"
+			var/cgstr = "CG"
+
+			// GUARD CHECK - Only search occupant for this specific ref, since your
+			//  can only CRISPR existing mutations in a target
+			var/datum/mutation/human/targetmut = get_mut_by_ref(bref, search_flags)
+			/* DEBUG SECTION: Dump Strings:
```
Comment (stylemistake) [inline code/game/machinery/computer/dna_console.dm:609]: This should not remain in the final PR.

File: code/game/machinery/computer/dna_console.dm [hunk 3]
```diff
@@ -549,6 +554,135 @@
...
+				for(var/i=1 to length(truegenes))
+					newString = (i%2==0 ? truegenes[i-1] : sequence[i])
+				to_chat(usr,"<span class='notice'>"+HM.name+" : "+truegenes+" : "+newString+"</span>")
+			to_chat(usr,"<span class='warning'>"+targetmut.name+" : "+sequence+"</span>")
+			*/
+
+			// Prompt for modifier string
+			var/newSequenceInput = input(usr, "Enter replacement sequence (or nothing to cancel)", "Replace inherent gene","")
+			// Drop out if the string is the wrong length
+			if(!(length(newSequenceInput)==32))
```
Comment (stylemistake) [inline code/game/machinery/computer/dna_console.dm:623]:
```suggestion
			if(length(newSequenceInput) != 32)
```

File: code/game/machinery/computer/dna_console.dm [hunk 4]
```diff
@@ -549,6 +554,135 @@
...
+				newpair = (pairstr ? char + (pairstr[1]==char?pairstr[2]:pairstr[1]) : null)
+				// every secong letter in the sequence represents a valid pair of the new sequence, otherwise it belongs to old
+				if(newpair)
+					if(i%2==0)
+						newSequence+=newpair
+					else
+						oldSequence+=newpair
+				else
+					return //drop out, no pair
+			//logging for debug
```
Comment (stylemistake) [inline code/game/machinery/computer/dna_console.dm:648]: This should not remain in the final PR.

**Comment by read-0nly** [comment]
Thanks for reviewing! I'll make the suggested modifications as soon as I can

**Comment by stylemistake** [comment]
Looks ok and functional to me, didn't test though.

**Comment by Timberpoes** [comment]
Since I have in-depth knowledge and experience with Genetics in general, I'm going to do a quick testing sweep on this code to make sure everything works as expected.

File: tgui/packages/tgui/interfaces/DnaConsole.js [hunk 5]
```diff
@@ -653,6 +653,14 @@ const MutationInfo = (props, context) => {
...
                 source: mutation.Source,
               })} />
+            <Button
+              icon="syringe"
+              disabled={!mutation.Active}
+              content="CRISPR"
+              onClick={() => act('crispr', {
+                mutref: mutation.ByondRef,
+                source: mutation.Source,
+              })} />
```
Comment (Timberpoes) [inline tgui/packages/tgui/interfaces/DnaConsole.js:663]: Should also be something somewhere on the interface indicating how many CRISPR charges are available.

**Review by Timberpoes — CHANGES_REQUESTED** [review]

File: code/game/machinery/computer/dna_console.dm [hunk 6]
```diff
@@ -129,6 +129,9 @@
 	/// State of tgui view, i.e. which tab is currently active, or which genome we're currently looking at.
 	var/list/list/tgui_view_state = list()
 
+	///Counter for CRISPR charges
+	var/CRISPRcharges = 0
```
[the remark this answers is not in the document]
Comment (read-0nly) [reply code/game/machinery/computer/dna_console.dm:133]: Thanks I'll get it corrected as soon as I ca

**Comment by Timberpoes** [comment]
Code works as advertised - Although the functionality is not exactly intuitive without documentation.

Definitely liaise with the wiki gurus over in the #wiki-general on the official /tg/Station13 Discord if/when this gets merged and get the new functionality documented in the wiki.

File: tgui/packages/tgui/interfaces/DnaConsole.js [hunk 7]
```diff
@@ -653,6 +653,14 @@ const MutationInfo = (props, context) => {
                 is_activator: 0,
                 source: mutation.Source,
               })} />
+            <Button
+              icon="syringe"
+              disabled={!mutation.Active}
```
[the remark this answers is not in the document]
Comment (read-0nly) [reply tgui/packages/tgui/interfaces/DnaConsole.js:658]: Not sure how to check for that in TGUI - if you had some resources I could look over I'm happy to do my own reading otherwise I'll tinker tonight to figure it out

File: tgui/packages/tgui/interfaces/DnaConsole.js (hunk 5 above)
Comment (Timberpoes) [reply tgui/packages/tgui/interfaces/DnaConsole.js:663]: This will probably help with the other code review regarding disabling the CRISPR button.

In `/obj/machinery/computer/scan_consolenew/ui_data(mob/user)`, `var/list/data` is an associated list of data that is sent to the tgui interface.

You can set a value in the data list probably below `data["isScannerConnected"] = can_use_scanner` sending the number of CRISPR charges to tgui, which you can then grab and use as necessary.

If you have any questions, hit up the #coding-general on the /tg/Station13 discord. There's a few tgui wizards who can help you.

Comment (Timberpoes) [reply tgui/packages/tgui/interfaces/DnaConsole.js:663]: >Not sure how to check for that in TGUI - if you had some resources I could look over I'm happy to do my own reading otherwise I'll tinker tonight to figure it out

See above. Also https://github.com/tgstation/tgstation/blob/master/tgui/README.md and https://github.com/tgstation/tgstation/blob/master/tgui/docs/tutorial-and-examples.md

**Comment by read-0nly** [comment]
Should be good now! 
- Refactored all the vars image
- added comments
- added some sort of probability for the random disease to properly be an analog for the CRISPR charge going rogue
- and now the button only activates if both the mutation is solved and a CRISPR charge is available

**Comment by read-0nly** [comment]
I can't see a way to save a page as a draft on the wiki without just saving the page but if/when this gets merged, for the wiki:

```

==== CRISPR Editing ====

This allows you to swap out base mutations in a targeted way. Base mutations don't suffer from instability

===== Getting a CRISPR Charge =====
First, you'll need to collect a CRISPR charge. This is a sample of a virus with genetic abilities you can use to swap out base mutations.
* Get the Viro to make a virus with either [[Infections#Symptoms_Table|Dormant DNA Activator]] or [[Infections#Symptoms_Table|Viral Evolutionary Acceleration]] (Ideally otherwise benign)
* Have a containment protocol, get yourself scanned for diseases if you get symptoms, hope the cure is available if things go south
* Infect a contained monkey, use an Activator (not a Mutator) on it - it collects the charge when it collects chromosomes
* Feed the used activator to the console to add the charge to the console


===== Using a CRISPR Charge =====
You'll need the top row of the sequence pairs of both the mutation you want to target and the mutation you want to replace it with. Here's the basic flow:
* Solve the mutation you want or just use a mutator on a humanized monkey
* Take note of the top row
* On the subject you want to swap mutations on, solve the mutation you want to swap
* Use a CRISPR charge on it
* Feed in a special CRISPR string instructing the virus to replace that mutation with the desired mutation

====== Building the CRISPR String ======
For example, let's say Epilepsy is
* '''<code>A T T A C G C G A T T A C G C G</code>'''
* '''<code>T A A T G C G C T A A T G C G C</code>'''

And Telekinesis is 
* '''<code>A T A T C C G G A T T A C C G G</code>'''
* '''<code>T A T A G G C C T A A T G G C C</code>'''

Take note of the first row of both
* Epilepsy is '''<code>A T T A C G C G A T T A C G C G</code>'''
* Telekinesis is '''<code>A T A T C C G G A T T A C C G G</code>'''

You then need to weave these rows together, old-new-old-new-old-new like this:
* '''<code>_A T A T C C G G A T T A C C G G :NEW (Telekinesis)</code>'''
* '''<code>AATTTAATCCGCCGGGAATTTTAACCGCCGGG :CRISPR String</code>'''
* '''<code>A T T A C G C G A T T A C G C G_ :OLD (Epilepsy)</code>'''

Now, when you use CRISPR on the solved base mutation you intend to swap, feed in that string! It swaps Epilepsy out for Telekinesis 
* '''<code> AATTTAATCCGCCGGGAATTTTAACCGCCGGG </code>'''

If your string is not the right length, nothing happens, but if the string is wrong, you may end up with a Acid Flesh instead, but scrambled and with no guides - hope you've got activators! Careful not to be wrong multiple times, lest you overwrite multiple mutations to the same one and need to shuffle to get them back. Also the CRISPR charge, being a repurposed virus, might sometimes go rogue and infect you randomly.
```

File: code/game/machinery/computer/dna_console.dm [hunk 8]
```diff
@@ -549,6 +560,143 @@
...
+					return
+
+			//Currently selected mutation
+			var/bref = params["mutref"]
+
+			//Valid gene-pairs
+			var/at_str = "AT"
+			var/cg_str = "CG"
+
+			// GUARD CHECK - Only search occupant for this specific ref, since your
```
Comment (Timberpoes) [inline code/game/machinery/computer/dna_console.dm:612]:
```suggestion
			// GUARD CHECK - Only search occupant for this specific ref, since you
```

File: code/game/machinery/computer/dna_console.dm [hunk 9]
```diff
@@ -96,6 +96,8 @@
 	var/is_joker_ready = FALSE
 	/// Used for setting tgui data - Whether injectors are ready to be printed
 	var/is_injector_ready = FALSE
+	/// USed for setting tgui data - Is CRISPR ready?
```
Comment (Timberpoes) [inline code/game/machinery/computer/dna_console.dm:99]:
```suggestion
	/// Used for setting tgui data - Is CRISPR ready?
```

File: code/game/machinery/computer/dna_console.dm [hunk 10]
```diff
@@ -549,6 +560,143 @@
...
+			//Unzip the modification string
+			for(var/i = 1 to length(new_sequence_input))
+				var/char = new_sequence_input[i]
+				var/pair_str
+				var/new_pair
+				//figure out which pair type the character belongs to
+				pair_str = ((at_str[1] == char || at_str[2] == char) ? at_str : ((cg_str[1] == char || cg_str[2] == char) ? cg_str : null))
+				//Valid pair from character
+				new_pair = (pair_str ? char + (pair_str[1]==char?pair_str[2]:pair_str[1]) : null)
+				// every secong letter in the sequence represents a valid pair of the new sequence, otherwise it belongs to old
```
Comment (Timberpoes) [inline code/game/machinery/computer/dna_console.dm:636]:
```suggestion
				// every second letter in the sequence represents a valid pair of the new sequence, otherwise it belongs to old
```

**Review by Timberpoes — CHANGES_REQUESTED** [review]
Proof reading comments.

**Comment by Timberpoes** [comment]
Minus a few comment typos, everything seems good to me now. Good luck and I hope you manage to get this merged!

**Comment by Jared-Fogle** [comment]
CRISPR is almost certainly trademarked. Can you come up with an original name?

**Comment by fludd12** [comment]
Obligatory "I Am Not A Lawyer" but CRISPR is a family of DNA sequences, and as far as I'm aware it can't be intellectual property -- the actual process required to modify DNA with it can be patented, but we're not replicating that here. It might be worth doing some research into it just to make sure we can actually use it by name, though.

**Comment by read-0nly** [comment]
This game has so many ripoffs that I honestly can't tell if that's a serious criticism, but CRISPR (or rather CRISPR-Cas9, for the context of gene editing) is the name of a process, not a trademark

**Comment by read-0nly** [comment]
[CRISPR (/ˈkrɪspər/) (clustered regularly interspaced short palindromic repeats) is a family of DNA sequences found in the genomes of prokaryotic organisms such as bacteria and archaea.](https://en.wikipedia.org/wiki/CRISPR)

[On June 13, 2013, in the case of the Association for Molecular Pathology v. Myriad Genetics, Inc., the Supreme Court of the United States ruled that human genes cannot be patented in the U.S. because DNA is a "product of nature."](https://ghr.nlm.nih.gov/primer/testing/genepatents)

**Comment by read-0nly** [comment]
Oh neat! Edited the wiki page to go with it!
